### PR TITLE
Issue #18646 Cleanup auth data from session if findIdentity() returns…

### PR DESCRIPTION
**Pull request by mikehaertl**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌
| Tests pass?   | ✔️
| Fixed issues  | #18646

### CODE REVIEW

File: framework/web/User.php [hunk 1]
```diff
@@ -700,18 +701,21 @@ protected function renewAuthStatus()
...
-            $authKey = $session->get($this->authKeyParam);
-            if ($authKey !== null && !$identity->validateAuthKey($authKey)) {
-                $identity = null;
-                $ip = Yii::$app->getRequest()->getUserIP();
-                Yii::warning("Invalid session auth key attempted for user '$id' from $ip: $authKey", __METHOD__);
-            }
+        if ($identity === null) {
+            $this->switchIdentity(null);
+            return;
+        }
```
Comment (mikehaertl) [inline framework/web/User.php:707]: @bizley I think you were right: We can do this here.

If we agree that `$identity` is the main source of truth, then that means: If there is no identity anymore (`findIdentity() === null`) then the user is no longer authenticated. We can stop any auto login attempts and don't need to check the auth key etc.

**Review by samdark — APPROVED** [review]

**Review by Arhell — APPROVED** [review]

**Comment by samdark** [comment]
Seems to fail current tests. @mikehaertl would you please check it?

**Comment by bizley** [comment]
@mikehaertl what is the status of this PR?

**Comment by mikehaertl** [comment]
@bizley Sorry, I've missed your question. From my side the PR should be fine. I don't know what is the reason why it was never merged. Now we have conflicts (which weren't there earlier). Should I fix them?

**Comment by mikehaertl** [comment]
@samdark I see the tests fail now. I think they worked fine when I wrote this back in may. From the output I can not really see which tests fail or why. Any hint?

**Comment by mikehaertl** [comment]
@samdark @bizley Ok, figured it out. My change here breaks cookie based login. As there are also conflicts and the fix needs some work should I open a fresh PR? If not, how do you deal with conflicts. In my projects I'd do a rebase on master. But I think you don't like that here?

**Comment by bizley** [comment]
Conflict resolved ;) it was just in the changelog.

**Comment by mikehaertl** [comment]
@bizley The test `tests/framework/filters/auth/BasicAuthTest::testHttpBasicAuthIssue15658` fails. I guess that the problem in that method is caused by [this](https://github.com/yiisoft/yii2/blob/c8e4e0727a122fddcd021f4419b939dc841d769c/tests/framework/filters/auth/BasicAuthTest.php#L112):

```
        $idValue = 'should not be changed';
        $session->set($idParam, $idValue);
```

This `$idValue` is not defined in `tests/framework/filters/stubs/UserIdentity::$ids`. Our change will delete invalid id data from session.

I'm unsure though how `tests/framework/filters/auth/AuthTest::ensureFilterApplies()` works, but it probably leads to a call of `User::getIsGuest()` and this will trigger `renewAuthStatus()` which removes the data.

As a fix we could change the `$idValue` in this test to a valid id - but this could break the idea of that test method (which I don't fully understand yet).

**Comment by mikehaertl** [comment]
Honestly I think the failing test method is not optimal. This logic looks very questionable to me:

https://github.com/yiisoft/yii2/blob/c8e4e0727a122fddcd021f4419b939dc841d769c/tests/framework/filters/auth/BasicAuthTest.php#L105-L113

* `login()` in line 113 is called with all test users from `AuthTest::tokenProvider()`
* So some logins will succeed (user1-3), others will fail (unknown, null)
* Line 113 messes with the internals of the session auth state:  even if the login failed it will write auth data to the session

In any case line 113 creates a mismatch between the currently logged in (or not logged in) user and the Id stored in session.

This leaves the auth status in an invalid state! So not sure what the idea behind this test logic is.

Can someone clarify?

**Comment by samdark** [comment]
Yeah, the test is quite interesting. The test is for https://github.com/yiisoft/yii2/issues/15658 and, indeed, is quite cryptic. Not sure it can be more straightforward though via using only public API. But that's definitely desired.

**Comment by mikehaertl** [comment]
@samdark All that is really required to test #15658 is a logged in user. So it should be sufficient to change the call to `login()` to a fixed user like `user1` instead of `$user`.

Currently the login will fail for 2 users (`unknown` and `null`). So the author "fixed" this by hacking the `idParam` into the session and by filtering our the `null` user inside the `auth` callback. 

If you agree I would rewrite that test a little.

**Comment by mikehaertl** [comment]
@samdark Rewrote the failing test. This should be ready for review & merge now.

**Review by bizley — APPROVED** [review]

**Review by samdark — APPROVED** [review]

**Comment by bizley** [comment]
Great, thank you.

**Pull request merged**